File: Week 2/Activities/minichallenge 1 solution/courseworks/scripts/signal_generator.py
#!/usr/bin/env python
import rospy
import numpy as np
from std_msgs.msg import Float32

if __name__=='__main__':
    signal_pub=rospy.Publisher("signal",Float32, queue_size=10)
    time_pub=rospy.Publisher("time",Float32, queue_size=10)
    rospy.init_node("signal_generator")
    rate = rospy.Rate(10)
    init_time = rospy.get_time()

    while not rospy.is_shutdown():
        time = rospy.get_time()-init_time
        signal = np.sin(time)
        time_pub.publish(time)
        signal_pub.publish(signal)
        rospy.loginfo("The signal value is: %f at a time %f", signal, time)

        rate.sleep()

File: class_ws/src/courseworks/src/signal_generator.cpp
#include <ros/ros.h>
#include <std_msgs/Float32.h>

int main(int argc, char **argv){
    ros::init(argc, argv, "signal_generator");
    ros::NodeHandle nh;
    // Publishers for signal and time
    ros::Publisher pub_signal = nh.advertise<std_msgs::Float32>("/signal", 100);
    ros::Publisher pub_time = nh.advertise<std_msgs::Float32>("/time", 100);

    ros::Rate loop_rate(10);

    float time = 0.0;

    while(ros::ok()){
        std_msgs::Float32 msg_signal;
        std_msgs::Float32 msg_time;
        msg_signal.data = sin(time);
        msg_time.data = time;
        pub_signal.publish(msg_signal);
        pub_time.publish(msg_time);

        ROS_INFO("Signal: %f, Time: %f", msg_signal.data, msg_time.data);

        time += 0.1;
        loop_rate.sleep();
        ros::spinOnce();
    }
}